src/test/integration/integration-tests.py

```diff
@@ -0,0 +1,168 @@
+        test_invalid_get_attribute(self, "name", "Defend")
+
+        # Invalid keywords
+        test_invalid_get_attribute(self, "keywords", "block")
+
+        # energyMin out of range
+        config["validGetParams"]["energyMin"] = -1
+        minOutOfRange = auxillaries.get_cards(config["validGetParams"],
+                                              )
+        self.assertEqual(minOutOfRange.status_code, 400)
+        self.assertIn("energyMin or energyMax out of range",
+                      minOutOfRange.json()["developerMessage"])
+        config["validGetParams"]["energyMin"] = 0
+
+        # energyMax out of range
+        config["validGetParams"]["energyMax"] = 1000
+        maxOutOfRange = auxillaries.get_cards(config["validGetParams"],
+                                              )
+        self.assertEqual(maxOutOfRange.status_code, 400)
+        self.assertIn("energyMin or energyMax out of range",
+                      maxOutOfRange.json()["developerMessage"])
+        config["validGetParams"]["energyMax"] = 0
+
+    def test_post_card(self):
+        # Invalid body
+
+        # Invalid type
+        test_invalid_attribute(self, "type", "skill", False)
+
+        # Invalid color
+        test_invalid_attribute(self, "color", "red", False)
+
+        # Invalid rarity
+        test_invalid_attribute(self, "rarity", "basic", False)
+
+        # Invalid name
+        test_invalid_attribute(self, "name", "Defend", False)
+
+        # Invalid description
+        test_invalid_attribute(self, "description", "Gain 5 block.", False)
+
+        # Invalid energy
+        test_invalid_attribute(self, "energy", 1, False)
+
+    def test_put_card(self):
+        # Invalid ID
+        invalid_id = auxillaries.put_card(config["invalidCardId"],
+                                          config["validPostBody"])
+        self.assertEqual(invalid_id.status_code, 404)
+
+        # Invalid body
+
+        # Invalid type
+        test_invalid_attribute(self, "type", "skill", True)
+
+        # Invalid color
+        test_invalid_attribute(self, "color", "red", True)
+
+        # Invalid rarity
+        test_invalid_attribute(self, "rarity", "basic", True)
+
+        # Invalid name
+        test_invalid_attribute(self, "name", "Defend", True)
+
+        # Invalid description
+        test_invalid_attribute(self, "description", "Gain 5 block.", True)
+
+        # Invalid energy
+        test_invalid_attribute(self, "energy", 1, True)
+
+    # def test_delete_card:
+        # Invalid id
+
+
+# Test valid response in combined test
+def test_valid_response(self, response, code):
+    self.assertEqual(response.status_code, code)
+    self.assertIsNotNone(response.json()["data"])
+    self.assertEqual(response.json()["data"]["type"], "card")
+
+
+# Test invalid attribute in PUT or POST body
+def test_invalid_attribute(self, attribute, defaultValue, isPut):
+    config["validPostBody"]["data"]["attributes"][attribute] = (
+        "bad" + attribute + "()"
```

Comment:
I couldn't recall where did I see an article talking about using concatenation vs. string formatting in the first place. Long story short, although concatenation is slightly faster than string formatting in Python, many people (including me) still prefer using string formatting for the readability and avoiding the confusion. What do you think? I am pretty fine if you prefer the current method here.
* https://stackoverflow.com/questions/376461/string-concatenation-vs-string-substitution-in-python
* https://stackoverflow.com/questions/38722105/format-strings-vs-concatenation

Reply:
Interesting, I wasn't aware of the performance differences in the two methods. I think string formatting would make more sense here since there doesn't seem to be much of a need for performance gains in the integration tests since I don't think they will be run too frequently. The script is also fairly small and doesn't take much time as it is (about 0.75 seconds) so readability seems more important here.